lite2/client.go

```diff
@@ -0,0 +1,374 @@
+// intermediate headers will be requested. See the specification for the
+// algorithm.
+//
+// If the trusted header is more recent than one here, an error is returned.
+func (c *Client) VerifyHeader(newHeader *types.SignedHeader, newVals *types.ValidatorSet, now time.Time) error {
+	if c.trustedHeader.Height >= newHeader.Height {
+		return errors.Errorf("height #%d is already trusted (last: #%d)", newHeader.Height, c.trustedHeader.Height)
+	}
+
+	var err error
+	switch c.mode {
+	case sequential:
+		err = c.sequence(newHeader, newVals, now)
+	case skipping:
+		err = c.bisection(c.trustedHeader, c.trustedNextVals, newHeader, newVals, now)
+	}
+	if err != nil {
+		return err
+	}
+
+	// Update trusted header and vals.
+	nextVals, err := c.primary.ValidatorSet(newHeader.Height + 1)
+	if err != nil {
+		return err
+	}
+	return c.updateTrustedHeaderAndVals(newHeader, nextVals)
+}
+
+func (c *Client) sequence(newHeader *types.SignedHeader, newVals *types.ValidatorSet, now time.Time) error {
+	// 1) Verify any intermediate headers.
+	var (
+		interimHeader *types.SignedHeader
+		nextVals      *types.ValidatorSet
+		err           error
+	)
+	for height := c.trustedHeader.Height + 1; height < newHeader.Height; height++ {
+		interimHeader, err = c.primary.SignedHeader(height)
+		if err != nil {
+			return errors.Wrapf(err, "failed to obtain the header #%d", height)
+		}
+
+		err = Verify(c.chainID, c.trustedHeader, c.trustedNextVals, interimHeader, c.trustedNextVals, c.trustingPeriod, now, c.trustLevel)
+		if err != nil {
+			return errors.Wrapf(err, "failed to verify the header #%d", height)
+		}
+
+		// Update trusted header and vals.
+		if height == newHeader.Height-1 {
+			nextVals = newVals
+		} else {
+			nextVals, err = c.primary.ValidatorSet(height + 1)
+			if err != nil {
+				return errors.Wrapf(err, "failed to obtain the vals #%d", height+1)
+			}
+		}
+		if !bytes.Equal(interimHeader.NextValidatorsHash, nextVals.Hash()) {
+			return errors.Errorf("expected next validator's hash %X, but got %X (height #%d)",
+				interimHeader.NextValidatorsHash,
+				nextVals.Hash(),
+				height)
+		}
+		err = c.updateTrustedHeaderAndVals(interimHeader, nextVals)
+		if err != nil {
+			return errors.Wrapf(err, "failed to update trusted state #%d", height)
+		}
+	}
+
+	// 2) Verify the new header.
+	return Verify(c.chainID, c.trustedHeader, c.trustedNextVals, newHeader, newVals, c.trustingPeriod, now, c.trustLevel)
+}
+
+func (c *Client) bisection(
+	lastHeader *types.SignedHeader,
+	lastVals *types.ValidatorSet,
+	newHeader *types.SignedHeader,
+	newVals *types.ValidatorSet,
+	now time.Time) error {
+
+	err := Verify(c.chainID, lastHeader, lastVals, newHeader, newVals, c.trustingPeriod, now, c.trustLevel)
+	switch err.(type) {
+	case nil:
+		return nil
+	case types.ErrTooMuchChange:
+		// continue bisection
+	default:
+		return errors.Wrapf(err, "failed to verify the header #%d ", newHeader.Height)
+	}
+
+	if newHeader.Height == c.trustedHeader.Height+1 {
```

Comment:
I don't understand this. There *can* be more than 1/3 change between subsequent headers, but since we have the `NextValidatorSetHash` (signed over by the previous validator set), we can verify the subsequent header.

Reply:
```
  if h2.Header.height == h1.Header.height + 1 {
    // we have adjacent headers that are not matching (failed
    // the CheckSupport)
    // we could submit evidence here
    return false
  }
```

from the spec

Reply:
> There can be more than 1/3 change between subsequent headers, but since we have the NextValidatorSetHash (signed over by the previous validator set), we can verify the subsequent header.

if > 2/3 of new vals signed h2 and h2.ValidatorHash == h1NextVals, then Verify will return nil (no error). See Verify

Reply:
Ah I see, OK then!

Reply:
Ah I see, you are right, that is the control flow.